analytical_engine/apps/assortativity/degree/degree_assortativity_coefficient.h

```diff
@@ -0,0 +1,208 @@
+                << std::to_string(a.first.second) + ": " << a.second
+                << std::endl;
+      }
+      ctx.merge_stage = true;
+      if (frag.fid() != 0) {
+        messages.SendToFragment(0, ctx.degree_mixing_map);
+      }
+      messages.ForceContinue();
+    } else {
+      // merge in work 0
+      if (frag.fid() == 0) {
+        std::unordered_map<std::pair<int, int>, int> msg;
+        while (messages.GetMessage(msg)) {
+          for (auto& a : msg) {
+            // update max degree
+            ctx.max_degree =
+                ctx.max_degree > a.first.first ? ctx.max_degree : a.first.first;
+            ctx.max_degree = ctx.max_degree > a.first.second ? ctx.max_degree
+                                                             : a.first.second;
+            // merge
+            if (ctx.degree_mixing_map.count(a.first) != 0) {
+              ctx.degree_mixing_map[a.first] += a.second;
+            } else {
+              ctx.degree_mixing_map[a.first] = a.second;
+            }
+          }
+        }
+        std::vector<std::vector<double>> degree_mixing_matrix(
+            ctx.max_degree + 1, std::vector<double>(ctx.max_degree + 1, 0.0));
+        GetDegreeMixingMatrix(ctx, degree_mixing_matrix);
+        for(auto& vec : degree_mixing_matrix){
+          for(auto& a : vec){
+            VLOG(0) << "degree mixing matrix: " << a << std::endl;
+          }
+        }
+        ctx.degree_assortativity = ProcessMatrix(degree_mixing_matrix);
+        VLOG(0) << "degree assortatity: " << ctx.degree_assortativity << std::endl;
+        for (auto& a : ctx.degree_mixing_map) {
+          VLOG(0) << std::to_string(a.first.first) + "-"
+                  << std::to_string(a.first.second) + ": " << a.second
+                  << std::endl;
+        }
+      }
+    }
+  }
+  double ProcessMatrix(std::vector<std::vector<double>>& degree_mixing_matrix){
+    int n = degree_mixing_matrix.size();
+    std::vector<double> a;
+    // sum of column
+    for(auto& row : degree_mixing_matrix){
+      a.emplace_back(accumulate(row.begin(), row.end(), 0.0));
+    }
+    std::vector<double> b;
+    // sum of row
+    for(int i = 0; i < n; i++){
+      double sum = 0.0;
+      for(int j = 0; j < n; j++){
+        sum += degree_mixing_matrix[j][i];
+      }
+      b.emplace_back(sum);
+    }
+    double sum = 0.0;
+    for(int i = 0; i < n; i++){
+      for(int j = 0; j < n; j++){
+        sum += i * j * (degree_mixing_matrix[i][j] - a[i] * b[j]);
+      }
+    }
+    double vara = Variance(a);
+    double varb = Variance(b);
+    return sum / (vara * varb);
+  }
+
+  double Variance(std::vector<double>& vec){
+    double sum1 = 0.0, sum2 = 0.0;
+    int n = vec.size();
+    for(int i = 0; i < n; i++){
+      sum1 += i * i * vec[i];
+      sum2 += i * vec[i];
+    }
+    return sqrt(sum1 - sum2 * sum2);
+  }
+  int GetDegreeByType(const fragment_t& frag, const vertex_t& vertex, DegreeType type) {
```

Comment:
seems the app did not support undirected graph ? like NetworkX, we should consider both directed graph and undirected gaph.

Reply:
The app supports both directed graph and undirected graph.For undirected graph, the result of GetLocalInDegree and GetLocalOutDegree is same.